// ipfs-http-client/esm/src/dht/put.js
import { Multiaddr } from 'multiaddr';
import { objectToCamel } from '../lib/object-to-camel.js';
import { configure } from '../lib/configure.js';
import { toUrlSearchParams } from '../lib/to-url-search-params.js';
import { multipartRequest } from 'ipfs-core-utils/multipart-request';
import { abortSignal } from '../lib/abort-signal.js';
import { AbortController } from 'native-abort-controller';
import { toString as uint8ArrayToString } from 'uint8arrays/to-string';
export const createPut = configure(api => {
  async function* put(key, value, options = {}) {
    const controller = new AbortController();
    const signal = abortSignal(controller.signal, options.signal);
    const res = await api.post('dht/put', {
      signal,
      searchParams: toUrlSearchParams({
        arg: uint8ArrayToString(key),
        ...options
      }),
      ...await multipartRequest([value], controller, options.headers)
    });
    for await (let message of res.ndjson()) {
      message = objectToCamel(message);
      if (message.responses) {
        message.responses = message.responses.map(({ID, Addrs}) => ({
          id: ID,
          addrs: (Addrs || []).map(a => new Multiaddr(a))
        }));
      }
      yield message;
    }
  }
  return put;
});
